### DISCUSSION TITLE
Is there an elegant way to invoke multiple query parallel when used with the suspense option?

### DISCUSSION THREAD

**Discussion by p-iknow** [TanStack/query#3973] [Q&A] [answered]
## Reproduce & Reference 
- [repo](https://github.com/p-iknow/next-js-react-query-playground)
- [codesandbox](https://codesandbox.io/s/admiring-water-g524bv?file=/pages/react-query-suspense-parallel.tsx)

## Question
Is there an elegant way to invoke multiple query parallel when used with the suspense option?
I am achieving my goal by calling useQuery  without a suspense option first as below temporarily. But this doesn't seem like a very elegant way. [react-query-suspense-parallel - code sandbox](https://codesandbox.io/s/admiring-water-g524bv?file=/pages/react-query-suspense-parallel.tsx)
![image](https://user-images.githubusercontent.com/35516239/182364726-f184ab6b-a5f1-4fde-a9d8-22be665e2c0e.png)
![image](https://user-images.githubusercontent.com/35516239/182367183-fba9cceb-8e6e-4b47-bbe6-b94af39f430c.png)

## Background 

I didn't want to express the components for the error and loading status as below. [- react-query-without-suspense-option - code sandbox](https://codesandbox.io/s/react-query-with-suspense-g524bv?file=/pages/react-query-without-suspense.tsx)
![image](https://user-images.githubusercontent.com/35516239/182365327-91c8848a-a01a-4c34-a29e-20513b264d58.png)

I wanted to express the error and loading status more declaratively. Thus, this was achieved through AsyncBoundary, which combines Suspense and ErrorBoundary. - [react-query-suspense-sequential](https://codesandbox.io/s/admiring-water-g524bv?file=/pages/react-query-suspense-sequential.tsx)
![image](https://user-images.githubusercontent.com/35516239/182367824-fb2c95b7-5bb1-4eb3-9f8a-a4bfdfa4ae58.png)
![image](https://user-images.githubusercontent.com/35516239/182369553-ef90f801-aa81-4581-8309-c3a6475ec8cf.png)

Everything seemed to go my way. But there was a price for using the suspense option.  Since suspense internally throw promise, it blocks another api call until one api call is completed. So if I look at the network waterfall, I can see that the api is called sequentially as shown below.

![image](https://user-images.githubusercontent.com/35516239/182367028-c32b8fcd-6dc9-4f66-84ed-009d6fa550d3.png)

As a way to solve the above problem, I decided to run the useQuery without the suspense option on the root before the useQuery using the suspense option on the sub-component is executed.
![image](https://user-images.githubusercontent.com/35516239/182370824-6fe97037-47ab-4dbe-822c-4323efaaa189.png)
![image](https://user-images.githubusercontent.com/35516239/182367183-fba9cceb-8e6e-4b47-bbe6-b94af39f430c.png)
In this way, the api is called parallelly. However, this method seems to be a temporary measure and not an elegant one. Is there a better way to achieve my goal?

## Requirements
- You have to use the suspense option.
- It should be possible to call the api in parallel.

## What I read before ask 
- https://github.com/TanStack/query/issues/1523
- https://github.com/TanStack/query/issues/2923#issuecomment-966703628

**Comment by TkDodo** [comment, marked as answer]
> In this way, the api is called parallelly. However, this method seems to be a temporary measure and not an elegant one. Is there a better way to achieve my goal?

I would like to get support for `suspense` on `useQueries`, but it's a long standing, open issue. What I think would need to happen is:
- go through the useQueries array and try to find all queries that have suspense on
- trigger a fetch for all non-suspense queries, e.g. with queryClient.prefetchQueries
- throw a new Promise with Promise.allSettled that contains the promises of all the queries created with `observer.fetchOptimistic`

I think what should happen then is that react suspends until all suspense queries have finished, while the non-suspense queries just run normally next to it. If you want to contribute, please give this a try :)
